File: converters/node/src/validator.ts
/**
 * Validation utilities for JSON Schema and GraphQL SDL
 */

import Ajv from 'ajv';
import addFormats from 'ajv-formats';
import { parse, buildSchema } from 'graphql';
import { JsonSchema, ValidationResult, ValidationError } from './types.js';

/**
 * Validator for JSON Schema and GraphQL SDL
 */
export class Validator {
  private ajv: Ajv;

  constructor() {
    this.ajv = new Ajv({
      allErrors: true,
      strict: false,
      validateFormats: true,
    });
    addFormats(this.ajv);
  }

  /**
   * Validate JSON Schema structure and x-graphql-* extensions
   *
   * @param schema - JSON Schema to validate
   * @returns Validation result
   */
  public validateJsonSchema(schema: JsonSchema): ValidationResult {
    const errors: ValidationError[] = [];

    // Basic structure validation
    if (typeof schema !== 'object' || schema === null) {
      return {
        valid: false,
        errors: [{ message: 'Schema must be an object', code: 'INVALID_SCHEMA' }],
      };
    }

    // Validate x-graphql-type-name if present
    if (schema['x-graphql-type-name']) {
      const nameValidation = this.validateGraphQLName(schema['x-graphql-type-name']);
      if (!nameValidation.valid) {
        errors.push(...(nameValidation.errors || []));
      }
    }

    // Validate x-graphql-type if present
    if (schema['x-graphql-type']) {
      const typeValidation = this.validateGraphQLType(schema['x-graphql-type']);
      if (!typeValidation.valid) {
        errors.push(...(typeValidation.errors || []));
      }
    }

    // Validate properties recursively
    if (schema.properties) {
      for (const [propName, propSchema] of Object.entries(schema.properties)) {
        const propValidation = this.validateJsonSchema(propSchema as JsonSchema);
        if (!propValidation.valid) {
          errors.push(
            ...(propValidation.errors || []).map((err) => ({
              ...err,
              path: `properties.${propName}${err.path ? '.' + err.path : ''}`,
            }))
          );
        }
      }
    }

    // Validate directives if present
    if (schema['x-graphql-directives']) {
      if (!Array.isArray(schema['x-graphql-directives'])) {
        errors.push({
          message: 'x-graphql-directives must be an array',
          code: 'INVALID_DIRECTIVES',
        });
      } else {
        for (const directive of schema['x-graphql-directives']) {
          if (!directive.name || typeof directive.name !== 'string') {
            errors.push({
              message: 'Directive must have a name',
              code: 'INVALID_DIRECTIVE',
            });
          } else {
            const nameValidation = this.validateGraphQLName(directive.name);
            if (!nameValidation.valid) {
              errors.push(...(nameValidation.errors || []));
            }
          }
        }
      }
    }

    return {
      valid: errors.length === 0,
      errors: errors.length > 0 ? errors : undefined,
    };
  }

  /**
   * Validate GraphQL SDL syntax
   *
   * @param sdl - GraphQL SDL string
   * @returns Validation result
   */
  public validateGraphQLSdl(sdl: string): ValidationResult {
    try {
      // Try to parse the SDL
      parse(sdl);

      // Try to build a schema (more comprehensive validation)
      buildSchema(sdl);

      return { valid: true };
    } catch (error) {
      return {
        valid: false,
        errors: [
          {
            message: (error as Error).message,
            code: 'INVALID_SDL',
          },
        ],
      };
    }
  }

  /**
   * Validate a GraphQL name
   *
   * @param name - Name to validate
   * @returns Validation result
   */
  public validateGraphQLName(name: string): ValidationResult {
    // GraphQL names must match: /^[_A-Za-z][_0-9A-Za-z]*$/
    const nameRegex = /^[_A-Za-z][_0-9A-Za-z]*$/;

    if (!name) {
      return {
        valid: false,
        errors: [{ message: 'Name cannot be empty', code: 'EMPTY_NAME' }],
      };
    }

    if (!nameRegex.test(name)) {
      return {
        valid: false,
        errors: [
          {
            message: `Invalid GraphQL name '${name}': must match /^[_A-Za-z][_0-9A-Za-z]*$/`,
            code: 'INVALID_NAME',
          },
        ],
      };
    }

    // Check for reserved names (starting with __)
    if (name.startsWith('__')) {
      return {
        valid: false,
        errors: [
          {
            message: `Invalid GraphQL name '${name}': names starting with '__' are reserved`,
            code: 'RESERVED_NAME',
          },
        ],
      };
    }

    return { valid: true };
  }

  /**
   * Validate a GraphQL type reference
   *
   * @param type - Type string to validate (e.g., "String!", "[Int]")
   * @returns Validation result
   */
  public validateGraphQLType(type: string): ValidationResult {
    // Basic validation for GraphQL type syntax
    const typeRegex = /^(\[)?[_A-Za-z][_0-9A-Za-z]*!?(\])?!?$/;

    if (!type) {
      return {
        valid: false,
        errors: [{ message: 'Type cannot be empty', code: 'EMPTY_TYPE' }],
      };
    }

    if (!typeRegex.test(type)) {
      return {
        valid: false,
        errors: [
          {
            message: `Invalid GraphQL type '${type}'`,
            code: 'INVALID_TYPE',
          },
        ],
      };
    }

    return { valid: true };
  }

  /**
   * Validate federation field selection syntax
   *
   * @param fields - Fields selection string
   * @returns Validation result
   */
  public validateFederationFields(fields: string): ValidationResult {
    if (!fields || fields.trim() === '') {
      return {
        valid: false,
        errors: [{ message: 'Fields selection cannot be empty', code: 'EMPTY_FIELDS' }],
      };
    }

    // Basic validation - would need more sophisticated parsing for full support
    const validChars = /^[_A-Za-z0-9\s"{}(),]+$/;
    if (!validChars.test(fields)) {
      return {
        valid: false,
        errors: [
          {
            message: `Invalid characters in fields selection: ${fields}`,
            code: 'INVALID_FIELDS',
          },
        ],
      };
    }

    return { valid: true };
  }

  /**
   * Validate a URL
   *
   * @param url - URL string to validate
   * @returns Validation result
   */
  public validateUrl(url: string): ValidationResult {
    const urlRegex = /^https?:\/\/[^\s/$.?#].[^\s]*$/i;

    if (!url) {
      return {
        valid: false,
        errors: [{ message: 'URL cannot be empty', code: 'EMPTY_URL' }],
      };
    }

    if (!urlRegex.test(url)) {
      return {
        valid: false,
        errors: [
          {
            message: `Invalid URL format: ${url}`,
            code: 'INVALID_URL',
          },
        ],
      };
    }

    return { valid: true };
  }
}

File: converters/rust/src/validator.rs
//! Validation utilities for JSON Schema and GraphQL SDL

use crate::error::{ConversionError, Result};
use regex::Regex;
use serde_json::Value as JsonValue;
use std::sync::OnceLock;

/// Regex for valid GraphQL names: /^[_A-Za-z][_0-9A-Za-z]*$/
fn graphql_name_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"^[_A-Za-z][_0-9A-Za-z]*$").unwrap())
}

/// Regex for valid GraphQL type references (including lists and non-null)
fn graphql_type_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"^(\[)?[_A-Za-z][_0-9A-Za-z]*!?(\])?!?$").unwrap())
}

/// Regex for valid federation field selections: fields(id, "name")
fn federation_fields_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| {
        Regex::new(r#"^[_A-Za-z][_0-9A-Za-z]*(\s+[_A-Za-z][_0-9A-Za-z]*|\s+"[^"]+")*(,\s*[_A-Za-z][_0-9A-Za-z]*(\s+[_A-Za-z][_0-9A-Za-z]*|\s+"[^"]+"))*$"#).unwrap()
    })
}

/// Regex for valid URLs
fn url_regex() -> &'static Regex {
    static REGEX: OnceLock<Regex> = OnceLock::new();
    REGEX.get_or_init(|| Regex::new(r"^https?://[^\s/$.?#].[^\s]*$").unwrap())
}

/// Validate a GraphQL name
pub fn validate_graphql_name(name: &str) -> Result<()> {
    if name.is_empty() {
        return Err(ConversionError::invalid_name(name, "name cannot be empty"));
    }

    if !graphql_name_regex().is_match(name) {
        return Err(ConversionError::invalid_name(
            name,
            "name must match /^[_A-Za-z][_0-9A-Za-z]*$/",
        ));
    }

    // Check for GraphQL reserved names
    if is_reserved_name(name) {
        return Err(ConversionError::invalid_name(
            name,
            "name is reserved by GraphQL specification",
        ));
    }

    Ok(())
}

/// Check if a name is reserved by GraphQL
fn is_reserved_name(name: &str) -> bool {
    // Names starting with __ are reserved
    name.starts_with("__")
}

/// Validate a GraphQL type reference
pub fn validate_graphql_type(type_str: &str) -> Result<()> {
    if type_str.is_empty() {
        return Err(ConversionError::InvalidType(
            "type cannot be empty".to_string(),
        ));
    }

    if !graphql_type_regex().is_match(type_str) {
        return Err(ConversionError::InvalidType(format!(
            "invalid GraphQL type format: {}",
            type_str
        )));
    }

    Ok(())
}

/// Validate federation field selection syntax
pub fn validate_federation_fields(fields: &str) -> Result<()> {
    if fields.is_empty() {
        return Err(ConversionError::FederationError(
            "fields selection cannot be empty".to_string(),
        ));
    }

    // Basic validation - more complex parsing would be needed for full support
    if !fields.chars().all(|c| {
        c.is_alphanumeric() || c == '_' || c == ' ' || c == '"' || c == ',' || c == '{' || c == '}'
    }) {
        return Err(ConversionError::FederationError(format!(
            "invalid characters in fields selection: {}",
            fields
        )));
    }

    Ok(())
}

/// Validate a URL
pub fn validate_url(url: &str) -> Result<()> {
    if !url_regex().is_match(url) {
        return Err(ConversionError::ValidationError(format!(
            "invalid URL format: {}",
            url
        )));
    }
    Ok(())
}

/// Validate JSON Schema structure
pub fn validate_json_schema(schema: &JsonValue) -> Result<()> {
    let mut errors = Vec::new();

    // Check if it's an object
    if !schema.is_object() {
        return Err(ConversionError::InvalidJsonSchema(
            "schema must be an object".to_string(),
        ));
    }

    let obj = schema.as_object().unwrap();

    // Validate $schema if present
    if let Some(schema_version) = obj.get("$schema") {
        if let Some(version_str) = schema_version.as_str() {
            if !version_str.contains("json-schema.org") {
                errors.push(ConversionError::validation(format!(
                    "invalid $schema URL: {}",
                    version_str
                )));
            }
        }
    }

    // Validate x-graphql-type-name if present
    if let Some(type_name) = obj.get("x-graphql-type-name") {
        if let Some(name) = type_name.as_str() {
            if let Err(e) = validate_graphql_name(name) {
                errors.push(e);
            }
        } else {
            errors.push(ConversionError::invalid_extension(
                "x-graphql-type-name must be a string",
            ));
        }
    }

    // Validate x-graphql-type if present
    if let Some(gql_type) = obj.get("x-graphql-type") {
        if let Some(type_str) = gql_type.as_str() {
            if let Err(e) = validate_graphql_type(type_str) {
                errors.push(e);
            }
        } else {
            errors.push(ConversionError::invalid_extension(
                "x-graphql-type must be a string",
            ));
        }
    }

    // Validate properties recursively
    if let Some(properties) = obj.get("properties") {
        if let Some(props_obj) = properties.as_object() {
            for (prop_name, prop_schema) in props_obj {
                // Validate property name format
                if prop_name.is_empty() {
                    errors.push(ConversionError::validation(
                        "property name cannot be empty".to_string(),
                    ));
                }

                // Recursively validate nested schemas
                if let Err(e) = validate_json_schema(prop_schema) {
                    errors.push(e);
                }
            }
        }
    }

    // Validate items for arrays
    if let Some(items) = obj.get("items") {
        if let Err(e) = validate_json_schema(items) {
            errors.push(e);
        }
    }

    // Validate x-graphql-directives if present
    if let Some(directives) = obj.get("x-graphql-directives") {
        if let Some(dir_array) = directives.as_array() {
            for directive in dir_array {
                if let Err(e) = validate_directive(directive) {
                    errors.push(e);
                }
            }
        } else {
            errors.push(ConversionError::invalid_extension(
                "x-graphql-directives must be an array",
            ));
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(ConversionError::combine(errors))
    }
}

/// Validate a directive definition
fn validate_directive(directive: &JsonValue) -> Result<()> {
    let obj = directive.as_object().ok_or_else(|| {
        ConversionError::InvalidDirective("directive must be an object".to_string())
    })?;

    // Validate directive name
    if let Some(name) = obj.get("name") {
        if let Some(name_str) = name.as_str() {
            validate_graphql_name(name_str)?;
        } else {
            return Err(ConversionError::InvalidDirective(
                "directive name must be a string".to_string(),
            ));
        }
    } else {
        return Err(ConversionError::InvalidDirective(
            "directive must have a name".to_string(),
        ));
    }

    // Validate arguments if present
    if let Some(args) = obj.get("arguments") {
        if !args.is_object() {
            return Err(ConversionError::InvalidDirective(
                "directive arguments must be an object".to_string(),
            ));
        }
    }

    Ok(())
}

/// Validate GraphQL SDL syntax (basic validation)
pub fn validate_graphql_sdl(sdl: &str) -> Result<()> {
    if sdl.trim().is_empty() {
        return Err(ConversionError::InvalidGraphQLSdl(
            "SDL cannot be empty".to_string(),
        ));
    }

    // Basic syntax checks
    let lines: Vec<&str> = sdl.lines().collect();
    let mut errors = Vec::new();
    let mut brace_count = 0;
    let mut in_type_definition = false;

    for (line_num, line) in lines.iter().enumerate() {
        let trimmed = line.trim();

        // Skip empty lines and comments
        if trimmed.is_empty() || trimmed.starts_with('#') {
            continue;
        }

        // Count braces for matching
        brace_count += trimmed.matches('{').count() as i32;
        brace_count -= trimmed.matches('}').count() as i32;

        // Check for type definition keywords
        if trimmed.starts_with("type ")
            || trimmed.starts_with("interface ")
            || trimmed.starts_with("enum ")
            || trimmed.starts_with("union ")
            || trimmed.starts_with("input ")
            || trimmed.starts_with("scalar ")
        {
            in_type_definition = true;

            // Extract type name after keyword
            let parts: Vec<&str> = trimmed.split_whitespace().collect();
            if parts.len() >= 2 {
                let type_name = parts[1].trim_end_matches('{').trim();
                if let Err(e) = validate_graphql_name(type_name) {
                    errors.push(ConversionError::InvalidGraphQLSdl(format!(
                        "Line {}: {}",
                        line_num + 1,
                        e
                    )));
                }
            }
        }

        // Check for field syntax inside type definitions
        if in_type_definition && brace_count > 0 && trimmed.contains(':') {
            // Basic field validation: should have "name: Type" format
            if !trimmed.starts_with("type ")
                && !trimmed.starts_with("interface ")
                && !trimmed.starts_with("enum ")
                && !trimmed.starts_with("union ")
                && !trimmed.starts_with("input ")
            {
                let parts: Vec<&str> = trimmed.split(':').collect();
                if parts.len() < 2 {
                    errors.push(ConversionError::InvalidGraphQLSdl(format!(
                        "Line {}: Invalid field syntax",
                        line_num + 1
                    )));
                }
            }
        }

        // Check for text without colons inside type bodies (likely invalid)
        if in_type_definition
            && brace_count > 0
            && !trimmed.is_empty()
            && !trimmed.contains(':')
            && !trimmed.starts_with('@')
            && !trimmed.contains('{')
            && !trimmed.contains('}')
            && !trimmed.starts_with("\"")
            && !trimmed.starts_with("type ")
            && !trimmed.starts_with("interface ")
            && !trimmed.starts_with("enum ")
            && !trimmed.starts_with("union ")
            && !trimmed.starts_with("input ")
        {
            // Check if it looks like it should be a field (has alphanumeric chars)
            if trimmed.chars().any(|c| c.is_alphanumeric()) {
                errors.push(ConversionError::InvalidGraphQLSdl(format!(
                    "Line {}: Invalid field syntax - missing colon or type",
                    line_num + 1
                )));
            }
        }

        if brace_count == 0 && in_type_definition {
            in_type_definition = false;
        }
    }

    // Check for unmatched braces
    if brace_count != 0 {
        errors.push(ConversionError::InvalidGraphQLSdl(
            "Unmatched braces in SDL".to_string(),
        ));
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(ConversionError::combine(errors))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_validate_graphql_name() {
        assert!(validate_graphql_name("User").is_ok());
        assert!(validate_graphql_name("_privateField").is_ok());
        assert!(validate_graphql_name("field123").is_ok());

        assert!(validate_graphql_name("").is_err());
        assert!(validate_graphql_name("123invalid").is_err());
        assert!(validate_graphql_name("invalid-name").is_err());
        assert!(validate_graphql_name("__reserved").is_err());
    }

    #[test]
    fn test_validate_graphql_type() {
        assert!(validate_graphql_type("String").is_ok());
        assert!(validate_graphql_type("String!").is_ok());
        assert!(validate_graphql_type("[String]").is_ok());
        assert!(validate_graphql_type("[String!]!").is_ok());

        assert!(validate_graphql_type("").is_err());
        assert!(validate_graphql_type("123Invalid").is_err());
    }

    #[test]
    fn test_validate_url() {
        assert!(validate_url("https://example.com").is_ok());
        assert!(validate_url("http://localhost:8080/path").is_ok());

        assert!(validate_url("not-a-url").is_err());
        assert!(validate_url("ftp://example.com").is_err());
    }

    #[test]
    fn test_validate_json_schema() {
        let valid_schema = json!({
            "$schema": "https://json-schema.org/draft/2020-12/schema",
            "type": "object",
            "x-graphql-type-name": "User",
            "properties": {
                "id": {
                    "type": "string",
                    "x-graphql-type": "ID!"
                }
            }
        });

        assert!(validate_json_schema(&valid_schema).is_ok());

        let invalid_schema = json!({
            "type": "object",
            "x-graphql-type-name": "123Invalid"
        });

        assert!(validate_json_schema(&invalid_schema).is_err());
    }

    #[test]
    fn test_validate_directive() {
        let valid_directive = json!({
            "name": "key",
            "arguments": {
                "fields": "id"
            }
        });

        assert!(validate_directive(&valid_directive).is_ok());

        let invalid_directive = json!({
            "arguments": {}
        });

        assert!(validate_directive(&invalid_directive).is_err());
    }

    #[test]
    fn test_validate_graphql_sdl() {
        let valid_sdl = r#"
            type User {
                id: ID!
                name: String
            }
        "#;

        assert!(validate_graphql_sdl(valid_sdl).is_ok());

        let invalid_sdl = "";
        assert!(validate_graphql_sdl(invalid_sdl).is_err());
    }

    #[test]
    fn test_reserved_names() {
        assert!(is_reserved_name("__typename"));
        assert!(is_reserved_name("__schema"));
        assert!(!is_reserved_name("_privateField"));
        assert!(!is_reserved_name("User"));
    }
}
